systemd-boot: allow configuration of log level

FelixPehla opened the pull request:
As discussed in #38678, here is a PR for introducing log levels into systemd-boot. It allows configuring a log level by using an SMBIOS type 11 string, or use `EFI_LOG_INFO` if none is specified. Depending on the feedback to this PR I'd also like to allow configuring via an option in `loader.conf`, or via efivars.

It is a Draft because there are still some things I'd like some input on:

 * previously, only `log_{debug,info,warning,warning_status,error,error_status,oom}(...)` existed, which is why I only used 4 log levels. Is having the standard log levels preferred, or do the existing ones provide enough granularity?
 * I'm using `#define` for specifying log levels to keep the semantics somewhat similar to syslog, but if that doesn't matter, using an `enum` would probably be nicer.
 * While reading the log level from an SMBIOS type 11 string can be done very early in the boot process, other options might set it late enough that some desired logs may be lost. This could be handled by keeping logs around until a level is explicitly configured (which might be bad for memory usage), or the log level could be saved as an efivar (and possibly configured by `bootctl`). Thoughts?

Open TODOs:
 * documentation (`smbios-type-11(7)`, `systemd-boot(7)`, etc.)
 * revert #38678 if this gets accepted

bluca commented:
> * previously, only `log_{debug,info,warning,warning_status,error,error_status,oom}(...)` existed, which is why I only used 4 log levels. Is having the standard log levels preferred, or do the existing ones provide enough granularity?

Just map level lower than err to err, so that the usual parameters still work

poettering commented on src/boot/boot.c:
```diff
@@ -2974,6 +2974,19 @@ static EFI_STATUS discover_root_dir(EFI_LOADED_IMAGE_PROTOCOL *loaded_image, EFI
+        if (is_confidential_vm())
+                return; /* Don't consume SMBIOS in CoCo contexts */
+
+        const char *level_str = smbios_find_oem_string("io.systemd.boot.loglevel=", /* after= */ NULL);
```
please document this in the man pages, i.e. systemd-boot(7) and smbios-type-11(7)

poettering commented:
looks pretty good, but let's try to replicate the bsd log levels here (both in strings and numerically), even if we won#t use the higher levels. It just makes things easier in the long run. for example we could eventually add something similar for systemd-stub that also configures the kernel's log level by adding something to the kernel cmdline, and then it gets really annoying if the set of levels is different...

(frnakly, we really should add this for systemd-stub too as suggested, that#d be supper nice i we had a single smbios knob that enables the log level from beginning to end.)

yuwata commented on src/boot/efi-log.c:
```diff
@@ -6,6 +6,17 @@
 
 static unsigned log_count = 0;
 
+static const uint8_t log_level_color[] = {
```
Please set `_LOG_MAX` for safety:
```c
static const uint8_t log_level_color[_LOG_MAX] = {
```

yuwata commented on src/boot/efi-log.c:
```diff
@@ -17,6 +19,39 @@ static const uint8_t log_level_color[] = {
         [LOG_DEBUG]   = EFI_LIGHTGRAY,
 };
 
+static const char *const log_level_table[] = {
```
Also here, please set `_LOG_MAX`.

yuwata commented on src/boot/efi-log.c:
```diff
@@ -17,6 +19,39 @@ static const uint8_t log_level_color[] = {
+        return old;
+}
+
+int log_set_max_level_from_string(const char *e) {
+        assert(e);
+        int r;
```
Please swap the ordering and insert empty lines:
1. declare `int r;`
2. insert an empty line,
3. `assert(e)`,
4. insert an empty line again.

```c
int log_set_max_level_from_string(const char *e) {
        int r;

        assert(e);

        r = log_level_from_string(e);
        ...
```

yuwata commented on src/boot/efi-log.c:
```diff
@@ -13,6 +15,34 @@ static const uint8_t log_level_color[] = {
+                return r;
+
+        log_set_max_level(r);
+        return 0;
```
OK.

yuwata commented on src/boot/boot.c:
```diff
@@ -2974,6 +2974,21 @@ static EFI_STATUS discover_root_dir(EFI_LOADED_IMAGE_PROTOCOL *loaded_image, EFI
+
+        r = log_set_max_level_from_string(level_str);
+        if (r < 0)
+                log_warning_status(EFI_SUCCESS, "Failed to parse log level '%s', ignoring.", level_str);
```
I think it is better to introduce `log_warning()` and use it here.

yuwata commented on src/boot/efi-log.h:
```diff
@@ -21,17 +21,30 @@ __attribute__((no_stack_protector, noinline)) void __stack_chk_guard_init(void);
+_gnu_printf_(3, 4) EFI_STATUS log_internal(EFI_STATUS status, LogLevel log_level, const char *format, ...);
+#define log_full(status, log_level, format, ...)                        \
+        log_internal(status, log_level, "%s:%i@%s: " format, __FILE__, __LINE__, __func__, ##__VA_ARGS__)
+#define log_debug(...) log_full(EFI_SUCCESS, LOG_DEBUG, __VA_ARGS__)
+#define log_info(...) log_full(EFI_SUCCESS, LOG_INFO, __VA_ARGS__)
+#define log_warning_status(status, ...) log_full(status, LOG_WARNING, __VA_ARGS__)
+#define log_error_status(status, ...) log_full(status, LOG_ERR, __VA_ARGS__)
+#define log_error(...) log_full(EFI_INVALID_PARAMETER, LOG_ERR, __VA_ARGS__)
```
When you touch the code, could you define all variations, like we do in src/basic/log.h ?

FelixPehla replied:
I'm assuming `log_warning()` should use `EFI_SUCCESS`, and `log_emergency()` `EFI_INVALID_PARAMETER` (like `log_error()` does right now)?

yuwata reviewed (changes requested):
Looks much much better. Several further comments.

FelixPehla commented on man/loader.conf.xml:
```diff
@@ -459,6 +459,28 @@ sbvarsign --attr "${attr}" --key KEK.key --cert KEK.pem --output db.auth db db.e
+        <literal>err</literal>, <literal>warning</literal>, <literal>notice</literal>,
+        <literal>info</literal>, and <literal>debug</literal>.</para>
+
+        <para>If unspecified, <literal>info</literal> will be used, unless unless one has already been
```
whoops, I don't know how I missed that...

yuwata commented on man/loader.conf.xml:
```diff
@@ -459,6 +459,28 @@ sbvarsign --attr "${attr}" --key KEK.key --cert KEK.pem --output db.auth db db.e
+        </para>
+
+        <para>Note that the configured level will only be used after <filename>loader.conf</filename> has
+        been parsed, so entries generated before that point may still show up on screen.</para>
```
It sound better to me.

yuwata commented on src/boot/boot.c:
```diff
@@ -151,6 +151,7 @@ typedef struct {
         bool sysfail_occured;
         int64_t console_mode;
         int64_t console_mode_efivar;
+        LogLevel log_level;
```
Hm? This seems to be only set. Maybe this can be dropped?

FelixPehla replied:
Yes, indeed

yuwata commented on src/boot/efi-log.c:
```diff
@@ -42,6 +42,10 @@ int log_set_max_level(LogLevel level) {
         return old;
 }
 
+LogLevel log_get_max_level(void) {
```
I'd like to also move this above `log_set_max_level()`, to make the order consistent with the header, but not a matter.

FelixPehla commented:
Alright, here's another update.

In addition to the requested changes, I've also added a corresponding output in `print_status()` and added `_pure_` and `_const_` where applicable.

And another question: since `EINVAL` is not available, I've just been using `-1` in `enum LogLevel` and `_DEFINE_STRING_TABLE_LOOKUP_FROM_STRING()`. Is that acceptable?

yuwata reviewed (approved):
LGTM. Several minor nitpicks.

FelixPehla commented:
Thank you very much for all the feedback! I'm still pretty inexperienced, but I learned a lot and hopefully my next PR won't need that much hand holding.

yuwata reviewed (approved):
LGTM. Thank you! Great work!

Could you cleanup the cover letter of this PR and drop stroked sentences and TODOs. The cover letter will be used in the merge commit, so clear description of the PR should be written there.

FelixPehla commented:
Does it look good as it is now?

Though now that I read over everything again, I wonder if the default log level shouldn't be `LOG_NOTICE` or `LOG_WARNING` instead. When this PR started out there were only 4 levels, so `LOG_INFO` seemed fine, but if we are using standard ones we might want to use `log_info()` more liberally without having everything shown on screen by default.

`log_info()` is currently only used in two places (outside of tests), so it won't change current behavior a lot either...

yuwata commented:
> Does it look good as it is now?

Yes. Thanks.

> Though now that I read over everything again, I wonder if the default log level shouldn't be `LOG_NOTICE` or `LOG_WARNING` instead. When this PR started out there were only 4 levels, so `LOG_INFO` seemed fine, but if we are using standard ones we might want to use `log_info()` more liberally without having everything shown on screen by default.
> 
> `log_info()` is currently only used in two places (outside of tests), so it won't change current behavior a lot either...

I think it is good to default to LOG_INFO, and that's consistent with our userspace tools and services.
Let's adjust log level of each message later one-by-one when necessary.
